Source/Core/VideoBackends/Vulkan/VulkanLoader.cpp

```diff
@@ -13,94 +13,91 @@
-#undef VULKAN_DEVICE_ENTRY_POINT
-#undef VULKAN_INSTANCE_ENTRY_POINT
-#undef VULKAN_MODULE_ENTRY_POINT
+namespace
+{
+#if defined(LOADER_WINDOWS)
+using Module = HMODULE;
+const char* const search_lib_names[] = {"vulkan-1.dll"};
 
-namespace Vulkan
+Module LoadModule(const char* lib_name) noexcept
 {
-static void ResetVulkanLibraryFunctionPointers()
+  return LoadLibraryA(lib_name);
+}
+
+void UnloadModule(Module* module) noexcept
 {
-#define VULKAN_MODULE_ENTRY_POINT(name, required) name = nullptr;
-#define VULKAN_INSTANCE_ENTRY_POINT(name, required) name = nullptr;
-#define VULKAN_DEVICE_ENTRY_POINT(name, required) name = nullptr;
-#include "VideoBackends/Vulkan/VulkanEntryPoints.inl"
-#undef VULKAN_DEVICE_ENTRY_POINT
-#undef VULKAN_INSTANCE_ENTRY_POINT
-#undef VULKAN_MODULE_ENTRY_POINT
+  FreeLibrary(*module);
+  *module = nullptr;
 }
 
-#if defined(VK_USE_PLATFORM_WIN32_KHR)
+void* LoadModuleFunction(Module module, const char* name) noexcept
+{
+  return static_cast<void*>(GetProcAddress(module, name));
+}
 
-static HMODULE vulkan_module;
-static std::atomic_int vulkan_module_ref_count = {0};
+#elif defined(LOADER_UNIX)
+using Module = void*;
+const char* const search_lib_names[] = {"libvulkan.so.1", "libvulkan.so"};
 
-bool LoadVulkanLibrary()
+Module LoadModule(const char* lib_name) noexcept
 {
-  // Not thread safe if a second thread calls the loader whilst the first is still in-progress.
-  if (vulkan_module)
-  {
-    vulkan_module_ref_count++;
-    return true;
-  }
-
-  vulkan_module = LoadLibraryA("vulkan-1.dll");
-  if (!vulkan_module)
-  {
-    ERROR_LOG(VIDEO, "Failed to load vulkan-1.dll");
-    return false;
-  }
+  return dlopen(lib_name, RTLD_NOW);
+}
 
-  bool required_functions_missing = false;
-  auto LoadFunction = [&](FARPROC* func_ptr, const char* name, bool is_required) {
-    *func_ptr = GetProcAddress(vulkan_module, name);
-    if (!(*func_ptr) && is_required)
-    {
-      ERROR_LOG(VIDEO, "Vulkan: Failed to load required module function %s", name);
-      required_functions_missing = true;
-    }
-  };
+void UnloadModule(Module* module) noexcept
+{
+  dlclose(*module);
+  *module = nullptr;
+}
 
-#define VULKAN_MODULE_ENTRY_POINT(name, required)                                                  \
-  LoadFunction(reinterpret_cast<FARPROC*>(&name), #name, required);
-#include "VideoBackends/Vulkan/VulkanEntryPoints.inl"
-#undef VULKAN_MODULE_ENTRY_POINT
+void* LoadModuleFunction(Module module, const char* name) noexcept
+{
+  return dlsym(module, name);
+}
 
-  if (required_functions_missing)
-  {
-    ResetVulkanLibraryFunctionPointers();
-    FreeLibrary(vulkan_module);
-    vulkan_module = nullptr;
-    return false;
-  }
+#else
+using Module = void*;
+const char* const search_lib_names[] = {};
 
-  vulkan_module_ref_count++;
-  return true;
+Module LoadModule(const char* lib_name) noexcept
+{
+  return nullptr;
 }
 
-void UnloadVulkanLibrary()
+void UnloadModule(Module* module) noexcept
```

Comment:
For consistency with the load function, I feel like the parameter should not be a pointer type (when the Module type itself is a pointer).